torchrl/objectives/value/vtrace.py

```
@@ -4,55 +4,334 @@
-    rho_bar: Union[float, torch.Tensor],
+    rho_thresh: Union[float, torch.Tensor],
     log_pi: torch.Tensor,
     log_mu: torch.Tensor,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
-    rho = _c_val(log_pi, log_mu, rho_bar)
-    next_vals = torch.cat([vals[:, 1:], torch.zeros_like(vals[:, :1])], 1)
-    dv = rho * (rewards + gamma * next_vals - vals)
-    return dv, rho
+    clipped_rho = _c_val(log_pi, log_mu, rho_thresh)
+    deltas = clipped_rho * (rewards + gamma * next_vals - vals)
+    return deltas, clipped_rho
 
 
-def _vtrace(
-    rewards: torch.Tensor,
-    vals: torch.Tensor,
-    log_pi: torch.Tensor,
-    log_mu: torch.Tensor,
-    gamma: Union[torch.Tensor, float],
-    rho_bar: Union[float, torch.Tensor] = 1.0,
-    c_bar: Union[float, torch.Tensor] = 1.0,
+# @_transpose_time # TODO: is this needed?
+def vtrace_correction(
+        gamma: float,
+        log_pi: torch.Tensor,
+        log_mu: torch.Tensor,
+        state_value: torch.Tensor,
+        next_state_value: torch.Tensor,
+        reward: torch.Tensor,
+        done: torch.Tensor,
+        rho_thresh: Union[float, torch.Tensor] = 1.0,
+        c_thresh: Union[float, torch.Tensor] = 1.0,
+        time_dim: int = -2,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
-    T = vals.shape[1]
-    if not isinstance(gamma, torch.Tensor):
-        gamma = torch.full_like(vals, gamma)
-
-    dv, rho = _dv_val(rewards, vals, gamma, rho_bar, log_pi, log_mu)
-    c = _c_val(log_pi, log_mu, c_bar)
-
-    v_out = []
-    v_out.append(vals[:, -1] + dv[:, -1])
-    for t in range(T - 2, -1, -1):
-        _v_out = (
-            vals[:, t] + dv[:, t] + gamma[:, t] * c[:, t] * (v_out[-1] - vals[:, t + 1])
+    """Computes V-Trace off-policy actor critic targets.
+
+    Refer to "IMPALA: Scalable Distributed Deep-RL with Importance Weighted  Actor-Learner Architectures"
+    https://arxiv.org/abs/1802.01561 for more context.
+
+    Args:
+        gamma (scalar): exponential mean discount.
```

will this work with gamma as a tensor (per-value gamma)?
If not we should raise an error saying that it is not supported yet and we welcome issues and contributions etc.